API Key rotation for Grafana Workspace

veerapratapg opened the pull request:
### What does this PR do?

<!-- A brief description of the change being made with this pull request. -->
- Added a module to enable Grafana API Key rotation --> [grafana-key-rotation](https://github.com/veerapratapg/terraform-aws-observability-accelerator/tree/main/modules/grafana-key-rotation) 
- Made appropriate changes to the [eks-monitoring](https://github.com/veerapratapg/terraform-aws-observability-accelerator/tree/main/modules/eks-monitoring) module to retrieve the details of the resources created in the in this module to be used for the above key rotation module. 
- Modified the [existing-cluster-with-base-and-infra](https://github.com/aws-observability/terraform-aws-observability-accelerator/tree/main/examples/existing-cluster-with-base-and-infra) example to create create the grafana-key-rotation module. 


### Motivation

<!-- What inspired you to submit this pull request? -->
- The current _[existing-cluster-with-base-and-infra](https://github.com/aws-observability/terraform-aws-observability-accelerator/tree/main/examples/existing-cluster-with-base-and-infra)_ example does not rotate the expired Grafana Workspace API keys automatically. 
- Modifying the example to enable key rotation for External Secrets deployed to the EKS Cluster as part of this solution. 
- Steps provided in https://docs.aws.amazon.com/prometheus/latest/userguide/obs_accelerator.html




### More

- [X] Yes, I have tested the PR using my local account setup  (Provide any test evidence report under Additional Notes)
- [X] Yes, I ran `pre-commit run -a` with this PR
- [ ] Yes, I have added a new example under [examples](https://github.com/aws-observability/terraform-aws-observability-accelerator/tree/main/examples) to support my PR (when applicable)
- [ ] Yes, I have updated the [Pages](https://github.com/aws-observability/terraform-aws-observability-accelerator/tree/main/docs) for this feature

**Note**: Not all the PRs required examples and docs.

### For Moderators
- [ ] E2E Test successfully complete before merge?

### Additional Notes

<!-- Anything else we should know when reviewing? -->

veerapratapg commented:
Tagging reviewers --> @bonclay7 @lewinkedrs

lewinkedrs commented on modules/grafana-key-rotation/main.tf:
```diff
@@ -0,0 +1,184 @@
+resource "aws_lambda_function" "observability_accelerator_lambda" {
+  function_name = "${var.lambda_function_name}-${var.managed_grafana_workspace_id}-${random_string.random_string_resource.id}"
+  handler       = "lambda_function.lambda_handler"
+  runtime       = "python3.8"
```
3.8 has a deprecation date of this year already. Can you test if this works properly with one of the newer run times? This way we don't have to worry about bumping the version later when it is deprecated.

lewinkedrs commented on modules/grafana-key-rotation/main.tf:
```diff
@@ -0,0 +1,184 @@
+
+# Policy Attachment to associate the custom policy created below to Lambda Execution Role
+resource "aws_iam_policy_attachment" "lambda_execution_role_policy_attach" {
+  name       = "test-attachment"
```
can we name this something other than 'test-attachment' . Having the word 'test' in there makes it sound unfinished. Maybe call it like "policy-attachment"

lewinkedrs commented on modules/grafana-key-rotation/main.tf:
```diff
@@ -0,0 +1,184 @@
+    actions = ["kms:Encrypt"]
+    effect = "Allow"
+    resources = [var.kms_key_arn_ssm] 
+
```
remove blank whitespace line

lewinkedrs commented on modules/grafana-key-rotation/main.tf:
```diff
@@ -0,0 +1,184 @@
+  target {
+    arn      = aws_lambda_function.observability_accelerator_lambda.arn
+    role_arn = aws_iam_role.eventbridge_scheduler_role.arn
+
```
remove blank whitespace line

lewinkedrs commented on modules/grafana-key-rotation/main.tf:
```diff
@@ -0,0 +1,184 @@
+
+# Policy Attachment to associate the custom policy created below to EventBridge Execution Role
+resource "aws_iam_policy_attachment" "eventbridge_scheduler_role_policy_attach" {
+  name       = "test-attachment"
```
as before, let's not use the name 'test-attachment'

lewinkedrs reviewed (commented):
Looks good to me but I did leave some comments with a few small fixes. We also need a doc for this. Can you write a markdown file in the docs folder describing how to use this module? Do not worry about the folder structure or where in the website to place the doc, we can do that part. But in the doc give a brief intro to what this module does, how to enable it or disable it, and describe some of the configuration options available to the users. Thank you for the contribution though it is looking very nice!

veerapratapg commented:
Made the following changes to incorporate the suggestions from @lewinkedrs :
1. Added a README.md doc to grafana-key-rotation module with the requested information. 
2. Performed testing with the latest version of python runtime for Lambda, "3.12", and the solution is working as expected. 
3. Removed `test` from the names of policy attachment resources. 
4. Also removed white spaces from the mentioned files. 
5. Added variables for python runtime, event bridge scheduler expressions to provide additional flexibility for users.

lewinkedrs reviewed (approved):
LGTM

bonclay7 reviewed (changes requested):
Can you please address this change, re-run pre-commit, and test it once more, as I've fixed some conflicts in your branch. 
Thanks again for the work!

veerapratapg commented:
Made the changes suggested by @bonclay7 and removed `managed_grafana_workspace_id` from the SSM parameter name created through `external-secrets` to reduce the number of changes made.

veerapratapg commented on modules/eks-monitoring/outputs.tf:
```diff
@@ -13,6 +13,21 @@ output "adot_irsa_arn" {
+output "kms_key_arn_eks_monitoring" {
+  description = "Name of the SSM Parameter"
+  value       = module.external_secrets[0].kms_key_arn_ssm
+}
```
Made the suggested change and removed `_eks_monitoring` from output names.

veerapratapg commented:
Added Unit Tests for Lambda function in the `grafana-key-rotation` module in the directory `./modules/grafana-key-rotation/tests/unit/src`. Also added a ReadMe file in the same location with a quick overview of steps to run the Tests.

veerapratapg commented:
Closing this PR as Managed Grafana v9 and 10 introduced Service account and service account token (SAT) APIs, instead of API Keys.
